firmware/library/L1_Drivers/can.hpp

```diff
@@ -1,49 +1,986 @@
+    {
+        // 00011 = Start of Frame
+        // 00010 = ID28 ... ID21
+        // 00110 = ID20 ... ID18
+        // 00100 = SRTR Bit
+        // 00101 = IDE bit
+        // 00111 = ID17 ... 13
+        // 01111 = ID12 ... ID5
+        // 01110 = ID4 ... ID0
+        // 01100 = RTR Bit
+        // 01101 = Reserved Bit 1
+        // 01001 = Reserved Bit 0
+        // 01011 = Data Length Code
+        // 01010 = Data Field
+        // 01000 = CRC Sequence
+        // 11000 = CRC Delimiter
+        // 11001 = Acknowledge Slot
+        // 11011 = Acknowledge Delimiter
+        // 11010 = End of Frame
+        // 10010 = Intermission
+
+         interrupts[controller_].ICR \
+         = can_registers[controller_]->ICR;
+         return interrupts[controller_].flags.error_code_location;
+    }
+
+    void EnableBus() override
+    {
+        SetControllerMode(kReset, false);
+    }
+
+    void ProcessIrq()
+    {
+        TxMessage_t * message_ptr;
+        bool transmit_interrupt, receive_interrupt;
+
+        // All bits are cleared, except the receive flag,
+        // when you read from the ICR register.
+        CaptureRegisterData();
+
+        transmit_interrupt = interrupts[controller_].flags.tx_buffer_1_ready ||
+                             interrupts[controller_].flags.tx_buffer_2_ready ||
+                             interrupts[controller_].flags.tx_buffer_3_ready;
+
+        receive_interrupt = interrupts[controller_].flags.rx_buffer_full;
+
+        if (transmit_interrupt == true)
+        {
+            if (xQueueIsQueueEmptyFromISR(tx_message_ptr_queue_) == pdFALSE)
+            {
+                // dequeue message pointer
+                xQueueReceiveFromISR(tx_message_ptr_queue_, &message_ptr, 0);
+
+                // send it to the available buffer
+                if (interrupts[controller_].flags.tx_buffer_1_ready)
+                {
+                    WriteMessageToBuffer(message_ptr, kBuffer1);
+                }
+                else if (interrupts[controller_].flags.tx_buffer_2_ready)
+                {
+                    WriteMessageToBuffer(message_ptr, kBuffer2);
+                }
+                else if (interrupts[controller_].flags.tx_buffer_3_ready)
+                {
+                    WriteMessageToBuffer(message_ptr, kBuffer3);
+                }
+                else
+                {
+                    // will never occur
+                }
+            }
+        }
+
+        if (receive_interrupt == true)
+        {
+            ReadMessageFromBuffer();
+        }
+    }
+
+ private:
+    Controllers controller_;
+    BaudRates baud_rate_;
+
+    PinInterface & rd_;
+    PinInterface & td_;
+
+    Pin rd_pin_;
+    Pin td_pin_;
+
+    QueueHandle_t tx_message_ptr_queue_;
+    QueueHandle_t rx_message_queue_;
+
+    inline void WriteMessageToBuffer(const TxMessage_t * const kMessage,
+                                     TxBuffers buffer_number)
+    {
+        /*
```

Comment:
Move all of the comments out for this review.

Reply:
done